### ISSUE TITLE
primaryKey option does not seem to be working

### ISSUE THREAD

**Issue by angelxmoreno** [overlookmotel/sequelize-hierarchy#67]
given the model:
```js
var Model = sequelize.define('Product', {
    id: {
        type: DataTypes.BIGINT,
        primaryKey: true
    },
    asin: {
        type: DataTypes.STRING,
    },
    parent_asin: {
        type: DataTypes.STRING
    },
    title: {
        type: DataTypes.STRING
    }
}, {
    hierarchy:{
        primaryKey: 'asin',
        foreignKey: 'parent_asin',
        levelFieldType: DataTypes.STRING
    },
    tableName: 'products'
});
return Model;
```

I am getting an error:
```
Key columns "parent_asin" and "id" are of incompatible types: character varying and bigint.
```

Am I doing something wrong?

**Comment by overlookmotel** [comment]
This looks like a bug. I'll look into it. But I'm very busy at the moment so not sure when I'll get to it.

Can you try one thing in the meantime? Remove the line `levelFieldType: DataTypes.STRING`. That seems like a bad idea to me and the primary key and level field don't need to be of same type.

That's probably unrelated to the problem you're having but worth trying. Please let me know if that solves it or not.

**Comment by overlookmotel** [comment]
Hmm. I can't immediately replicate this.

What versions of `sequelize` and `sequelize-hierarchy` are you on please?

And do you get the error when you define the model, when you call `sequelize.sync()`, or later on when you try to query the table/insert rows into the table?

**Comment by overlookmotel** [comment]
Ah OK. I think I've worked it out. The `parent` field must relate to the primary key field of the model. So you can change your model definition to:

```js
var Model = sequelize.define('Product', {
    id: {
        type: DataTypes.BIGINT,
        primaryKey: true
    },
    asin: {
        type: DataTypes.STRING,
    },
    parent_id: {
        type: DataTypes.BIGINT
    },
    title: {
        type: DataTypes.STRING
    }
}, {
    hierarchy:{
        foreignKey: 'parent_id'
    },
    tableName: 'products'
});
```

(so `parent_id` refers to `id`)

Or:

```js
var Model = sequelize.define('Product', {
    asin: {
        type: DataTypes.STRING,
        primaryKey: true
    },
    parent_asin: {
        type: DataTypes.STRING
    },
    title: {
        type: DataTypes.STRING
    }
}, {
    hierarchy:{
        foreignKey: 'parent_asin'
    },
    tableName: 'products'
});
```

(so `asin` becomes the primary key of the model)

Either way, get rid of the line `levelFieldType: DataTypes.STRING`. I can't be sure it'll work if the level field is not a number type.

**Comment by overlookmotel** [comment]
I'll look into getting your original code example to work. But in the meantime, I hope one of the above solutions will work for you. Let me know how you get on.

**Comment by overlookmotel** [comment]
Hold tight. I think I've got a fix to make your original code work. Just testing it and if all good I'll release shortly.

**Comment by angelxmoreno** [comment]
wow, thanks! A few things to note, I can not have `asin` as the primary key because not all products have an `asin` attached to it. 

These are the versions in my package.json
```
"sequelize": "^3.21.0",
"sequelize-hierarchy": "^0.5.8"
```

**Comment by overlookmotel** [comment]
But you could have the parent field refer to `id` rather than `asin` no?

**Comment by overlookmotel** [comment]
But yeah, I think my fix will make your original code work. Still waiting for Travis to finish running the tests. It's slooooow.

**Comment by overlookmotel** [comment]
Fixed in version 0.7.1, just published to npm.

The only breaking changes between 0.5.8 and 0.7.1 are small and relate to using schemas in postgres, so unless you're doing that, you should be fine up upgrade to latest.

Thanks for bringing this issue to my attention.

**Comment by overlookmotel** [comment]
Please let me know if new version works for you...

**Comment by angelxmoreno** [comment]
I will test this tonight. I am using postgres but I can switch to MySQL

**Comment by overlookmotel** [comment]
Postgres should be fine anyway. The changes between 0.5.8 and 0.7.1 only make a difference if you're using schemas, which is a fairly niche feature. And even then the changes were very minor.

No need to switch to MySQL!

**Comment by angelxmoreno** [comment]
👍

**Comment by angelxmoreno** [comment]
@overlookmotel testing things now. With the only change being the version bump, I am getting the same results. Is it my understanding that parentId must always be an integer? I wanted to avoid the extra look up to map asin and parent_asin to their integer primary keys. If the module does not support this I can redesign my schema. I was just under the impression that the module would auto create the id in the hierarchy table based off of the fields types defined in the model. Is this not the case?

**Comment by overlookmotel** [comment]
Did you get rid of the line `levelFieldType: DataTypes.STRING`?

**Comment by angelxmoreno** [comment]
yes. This is the sql Sequelize is attempting to run but failing:
```
CREATE TABLE IF NOT EXISTS `amazon_products` (
    `id` BIGINT ,
    `asin` VARCHAR(255),
    `parent_asin` VARCHAR(255),
    `detail_page_uRL` TEXT,
    `created` DATETIME NOT NULL,
    `modified` DATETIME NOT NULL,
    `deleted` DATETIME,
    `hierarchy_level` INTEGER UNSIGNED,
    PRIMARY KEY (`id`), 
    FOREIGN KEY (`parent_asin`)
        REFERENCES `amazon_products` (`id`) //shouldn't this be (`asin`) ?
                ON DELETE RESTRICT
                ON UPDATE CASCADE
) ENGINE=InnoDB;
```

**Comment by overlookmotel** [comment]
Yes, you're right. I could have sworn the change I made fixed it. But now it's not working on my machine either.

But I unlikely to have time to work this out in near future, I'm afraid.

I'd suggest making the parent field refer to `id` rather than `asin`. That will work around the problem, and also doesn't require that every item has an `asin`.

Here's what it'd look like:

```js
var Model = sequelize.define('Product', {
    id: {
        type: DataTypes.BIGINT,
        primaryKey: true
    },
    asin: {
        type: DataTypes.STRING,
    },
    parent_id: {
        type: DataTypes.BIGINT
    },
    title: {
        type: DataTypes.STRING
    }
}, {
    hierarchy:{
        foreignKey: 'parent_id'
    },
    tableName: 'products'
});
```

**Comment by angelxmoreno** [comment]
Unfortunately, when I am given a `Product`, I do not know the `id`. What I am given is the `parent_asin`. I wanted to avoid having to do a look up call for the `id` and then set that to the `parent_id`. However, I did take the advice of making the `asin` primary and that seemed to make all my issues go away! My model now looks like this:
```js
var Model = sequelize.define('AmazonProduct', {
    asin: {
        type: DataTypes.STRING,
        primaryKey: true
    },
    parent_asin: {
        type: DataTypes.STRING
    },
    detail_page_url: {
        type: DataTypes.TEXT
    }
}, {
    hierarchy: {
        primaryKey: 'asin',
        foreignKey: 'parent_asin',
        levelFieldName: 'hierarchy_level',
        through: 'AmazonProductAncestor',
        throughTable: 'amazon_product_ancestors',
        labels: true
    },
    tableName: 'amazon_products',
    indexes: [
        {
            unique: true,
            fields: ['asin']
        },
        {
            unique: false,
            fields: ['parent_asin']
        }
    ]
});
```
This is a small inconvenience and the docs do not reflects this behavior. It looks like the module takes the current model's primaryKey over the option `primaryKey`. I would not be surprised if I removed the `primaryKey` option and still get the exact sql produced.

The module has some bugs and the PR to introduce primaryKey customization seems to not do a complete job. All in all, this is still a great module and I really appreciate your time on this. Please leave this open so that all future devs are aware of this issue and how to get around it. Since I will be using this module a lot, I might just get in there and make good things happen. 

Again, thank you for your attention to this issue and for leading me to a resolution.

**Comment by overlookmotel** [comment]
@angelxmoreno You are right the PR to introduce `primaryKey` option didn't do the complete job. But I thought the change I made yesterday filled in the gap.

However, I think the ultimate cause of this is a bug in `sequelize` itself.

In case you want to investigate:

The change I made which I thought would fix this issue (https://github.com/overlookmotel/sequelize-hierarchy/commit/31847291c4daf1e0a2bdb434a97903063bb78b70) added a `targetKey` property to the `.belongsTo()` call that create the relationship. According to the Sequelize docs, this should set up the relationship as intended.

But it appears it's not working properly in Sequelize. There's various issues about the subject e.g. https://github.com/sequelize/sequelize/issues/4264

Certainly, if you go back to Sequelize v3.9.0, it gets closer to working, so it appears there's been a regression bug in Sequelize.

**Comment by overlookmotel** [comment]
#### Conclusion for anyone reading

1. `options.primaryKey` works if you already have your tables created in the database.
2. It does not work if you use `sequelize.sync()` to create the tables.
3. Cause of (2) is apparently bug in Sequelize itself concerning `targetKey` option in associations

**Comment by overlookmotel** [comment]
@devlato You originally requested the `primaryKey` option and wrote the PR https://github.com/overlookmotel/sequelize-hierarchy/pull/12 to implement. Do you have any thoughts on this?

**Comment by devlato** [comment]
@overlookmotel No idea yet, but I can research a bit

**Comment by overlookmotel** [comment]
@devlato That'd be great if you could. I'm really tied up with other projects, and I don't use this feature myself, so realistically I'm not going to be able to get to it any time soon. But it's part of the documented API so it'd be ideal if it works.

NB Your PR got everything right for handling cases where tables are already in the database and you don't use `include` or accessors (e.g. `.getParents()`). But it didn't do anything to create the right tables/relationships and `sequelize.sync()` malfunctions. It's also my bad for not checking it properly before merging and writing some tests.

**Comment by overlookmotel** [comment]
I've made a couple more changes to `master` branch which (I *think*) should fix this issue if Sequelize worked as documented.

I've also published a branch [issue-67](https://github.com/overlookmotel/sequelize-hierarchy/tree/issue-67) with a failing test for this issue. Here's the test: https://github.com/overlookmotel/sequelize-hierarchy/blob/e18ff31d312c8291a3e238cb066fb85812eb94b7/test/issue67.test.js

I *think* the problem is that Sequelize is ignoring the `targetKey` option in `.belongsTo()` and `.belongsToMany()`. To resolve this issue, it may also require `targetKey` to be implemented on `.hasMany()` too.

If anyone has time to narrow this down and raise issues on Sequelize to fix the problem, that'd be great!

**Comment by overlookmotel** [comment]
This bug still exists with Sequelize v4.3.2.

**Comment by Aralizer** [comment]
Any updates regarding the issue?
I'm trying to define the primary key as some other attribute. The through table definition works fine and take the correct value to create the links, but when trying to query for descendents, the actual query shows an incorrect join using the actual primary key of the table rather then the one I defined.

**Comment by overlookmotel** [comment]
I'm afraid it looks like Sequelize still have not fixed the bug with `targetKey` option in associations.

I just ran the tests with latest version of Sequlize and it's still failing: https://travis-ci.org/overlookmotel/sequelize-hierarchy/builds/361267709

Nothing I can do until that bug is fixed. If you want to hurry this along, I'd suggest doing a little research into Sequelize's use of `targetKey` option in associations and raising and issue there.

Or, if I'm wrong, and there is a bug in sequelize-hierarchy, I'd be very happy to receive a PR. Sorry but I don't have time myself to fix this.

**Comment by Aralizer** [comment]
I did some minor attempts with sequelize-hierarchy sources, specifically with lib/modelExtends.js file.
Found that a change on line 86

`this.hasMany(this, {as: options.childrenAs, foreignKey: options.foreignKey, targetKey: options.primaryKey, onDelete: options.onDelete});`
solves the `getChildren` and `include: [{as: 'children' ...}]` queries if you change the `targetKey` to `sourceKey` like this:

`this.hasMany(this, {as: options.childrenAs, foreignKey: options.foreignKey, sourceKey: options.primaryKey, onDelete: options.onDelete});`

According the the [docs](http://docs.sequelizejs.com/manual/tutorial/associations.html#one-to-many-associations-hasmany-) this is more fitting the the `hasMany` relationship. 
But I didn't run any tests and don't know if this breaks anything else...

Saw you proposed incomplete fix and tried that too, didn't help any of the 'decedents' or 'ancestors' queries...
